### PR TITLE
Major improvement to OSM-based electricity grid (e.g. using relations, preserving substation locations)

### PR DESCRIPTION
**Pull request by bobbyxng**
## Changes proposed in this Pull Request
- Major improvement to the OSM-based electricity grid
- `build_osm_network` completely rewritten:
  - Use of OSM relations, where available and where a single linestring could be extracted to replace member ways
  - substantial speed improvements
  - Exact locations of original substations are preserved
  - Lines with identical geometries and voltages are aggregated (num_parallel is increased)
  - Lines overpassing "virtual nodes" (not actual substations as per OSM raw data) are merged, if their electric properties allow it (same voltage levels, same number of circuits)
  - Geometries are preserved. Method to extend line/link to the substation has been improved significantly. No "mess" inside of a substation cluster anymore.
  - Method to determine substation point improved, based on "Pole of Inaccessibility" (PoI) method developed by Garcia-Castellanos & Lombardo 2007, doi.org/10.1080/14702540801897809 which ensures that an interior point in non-convex shapes is always found.
  - Clustering resolution has been sharpened from 5000 meter radius to only 500 meters, while still ensuring topological connectivity.
  - Single bus for each voltage level in each substation. Buses of different voltages are arranged in geometric shape (three buses form a triangle, see screenshot)
  - Transformers now have a capacity s_nom based on the maximum of connected capacity connected to each bus
  - Resolves many issues related around cross-border capacity, especially if the substation was located close to a border. @lindnemi @toniseibold @JulianGeis @nworbmot  

image

image

image

## Checklist

- [X] I tested my contribution locally and it works as intended.
- Done here: https://github.com/PyPSA/pypsa-eur/pull/1387
- [X] Code and workflow changes are sufficiently documented.
- [X] Changed dependencies are added to `envs/environment.yaml`.
- [X] Changes in configuration options are added in `config/config.default.yaml`.
- [X] Changes in configuration options are documented in `doc/configtables/*.csv`.
- [X] Sources of newly added data are documented in `doc/data_sources.rst`.
- [X] A release note `doc/release_notes.rst` is added.

### CODE REVIEW

File: scripts/base_network.py [hunk 1]
```diff
@@ -456,7 +456,8 @@ def _set_electrical_parameters_transformers(transformers, config):
 
     ## Add transformer parameters
     transformers["x"] = config.get("x", 0.1)
-    transformers["s_nom"] = config.get("s_nom", 2000)
+    if "s_nom" not in transformers:
+        transformers["s_nom"] = config.get("s_nom", 2000)
```
Comment (fneum) [inline scripts/base_network.py:460]: What happens if some transformers have `s_nom`, but not all of them. Will missing values be filled?

File: scripts/build_gdp_pop_non_nuts3.py [hunk 2]
```diff
@@ -123,7 +123,9 @@ def calc_gdp_pop(country, regions, gdp_non_nuts3, pop_non_nuts3):
     if "snakemake" not in globals():
         from _helpers import mock_snakemake
 
-        snakemake = mock_snakemake("build_gdp_pop_non_nuts3")
+        snakemake = mock_snakemake(
+            "build_gdp_pop_non_nuts3", configfiles=["config/config.osm-raw.yaml"]
+        )
```
Comment (fneum) [inline scripts/build_gdp_pop_non_nuts3.py:128]: revert?

File: config/config.default.yaml [hunk 3]
```diff
@@ -84,7 +84,7 @@ co2_budget:
 
 # docs in https://pypsa-eur.readthedocs.io/en/latest/configuration.html#electricity
 electricity:
-  voltages: [200., 220., 300., 380., 500., 750.]
+  voltages: [220., 300., 330., 380., 400., 500., 750.]
```
Comment (fneum) [inline config/config.default.yaml:87]: Just a short note why 200 kV disappears for clarification would be good.

File: scripts/build_osm_network.py [hunk 4]
```diff
@@ -41,823 +57,1576 @@
...
+            "voltage", None
+        )  # Default to None if not present
+
+        # Extract the geometries for all nodes in the subgraph
+        geometry = [G.nodes[node].get("geometry", None) for node in subgraph.nodes()]
+        geometry = linemerge(geometry)  # Merge the geometries
+
+        # # list of all countries
+        # country = ';'.join([G.nodes[node].get('country', '') for node in subgraph.nodes()])
+        # country = ';'.join(sorted(set(country.split(';'))))
```
Comment (fneum) [inline scripts/build_osm_network.py:427]: could remove unused code?

File: scripts/build_osm_network.py [hunk 5]
```diff
@@ -41,823 +57,1576 @@
...
+    all_transformers = gpd.GeoDataFrame(
+        columns=TRANSFORMERS_COLUMNS + ["transformer_id"], crs=geo_crs
+    )
 
-    for p in points:
-        # execute split to all lines and store results
-        temp_list = [split(l, p) for l in list_linestrings]
-        # nest all geometries
-        list_linestrings = [lstring for tval in temp_list for lstring in tval.geoms]
+    for g_name, g_value in buses_all.groupby("station_id"):
```
Comment (fneum) [inline scripts/build_osm_network.py:1125]: *If this is slow,* could be converted to an apply function applied to each group.

**Review by fneum — COMMENTED** [review]
I didn't check every line in detail, but the structure looks good and comparable to the previous version. I have a series of small comments that are easy to address:

File: config/config.default.yaml (hunk 3 above)
Comment (bobbyxng) [reply config/config.default.yaml:87]: Yes, I double-checked the voltage levels being used in Europe. Both JAO Static Grid Model and the ENTSO-E map show that that 200 kV is not a standardised value in operation. I then checked the few lines in OSM that were labeled as 200 kV lines, of which some were actually 20 kV lines (with a typo, trailing zero). Now the underlying OSM input data (at least for Europe) does not contain 200 kV lines anymore.

- copied from https://github.com/PyPSA/pypsa-eur/pull/1387#discussion_r1830634924

File: scripts/build_osm_network.py (hunk 4 above)
Comment (bobbyxng) [reply scripts/build_osm_network.py:427]: done

File: scripts/build_osm_network.py (hunk 5 above)
Comment (bobbyxng) [reply scripts/build_osm_network.py:1125]: it is surprisingly fast, few seconds, so no need for now from my side :)

File: scripts/build_gdp_pop_non_nuts3.py (hunk 2 above)
Comment (bobbyxng) [reply scripts/build_gdp_pop_non_nuts3.py:128]: done

**Comment by bobbyxng** [comment]
Docstrings added

**Pull request merged**